cpp/daal/src/algorithms/tsne/tsne_gradient_descent_avx512_impl.i

```diff
@@ -0,0 +1,326 @@
+/* file: tsne_gradient_descent_avx512_impl.i */
+/*******************************************************************************
+* Copyright 2022 Intel Corporation
+*
+* Licensed under the Apache License, Version 2.0 (the "License");
+* you may not use this file except in compliance with the License.
+* You may obtain a copy of the License at
+*
+*     http://www.apache.org/licenses/LICENSE-2.0
+*
+* Unless required by applicable law or agreed to in writing, software
+* distributed under the License is distributed on an "AS IS" BASIS,
+* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+* See the License for the specific language governing permissions and
+* limitations under the License.
+*******************************************************************************/
+
+/*
+//++
+//  Parts of tSNE algorithm oprtimized for AVX512.
+//--
+*/
+
+#ifndef __TSNE_GRADIENT_DESCENT_AVX512_IMPL_I__
+#define __TSNE_GRADIENT_DESCENT_AVX512_IMPL_I__
+
+namespace daal
+{
+namespace algorithms
+{
+namespace internal
+{
+
+/* Partial template specialization of attractive kernel for single precision data and AVX512 ISA */
+template <bool DivComp, typename IdxType>
+struct AttractiveKernel<DivComp, IdxType, float, avx512>
+{
+    static services::Status impl(const float * val, const IdxType * col, const size_t * row, MemoryCtxType<IdxType, float, avx512> & mem,
+                                 float & zNorm, float & divergence, const IdxType N, const IdxType nnz, const IdxType nElements,
+                                 const float exaggeration)
+    {
+        const float multiplier = exaggeration * float(zNorm);
```

Comment:
Should we check for overflow?

Reply:
No, we don't. We check only for integer overflow.